### PR TITLE
dealpublisher: Fully validate deals before publishing

### PR DESCRIPTION
**Pull request by magik6k**
Should fix https://github.com/filecoin-project/lotus/issues/7206

### CODE REVIEW

**Comment by magik6k** [comment]
This validation seems to be pretty fast, on my mainnet miner:
```
2021-08-31T14:02:02.435+0200    INFO    storageadapter  storageadapter/dealpublisher.go:370     validating deal {"took": 0.015638289, "proposal": "bafy..."}
```

File: markets/storageadapter/dealpublisher.go [hunk 1]
```diff
@@ -315,6 +318,13 @@ func (p *DealPublisher) publishReady(ready []*pendingDeal) {
 // validateDeal checks that the deal proposal start epoch hasn't already
```
Comment (arajasek) [inline markets/storageadapter/dealpublisher.go:318]: update comment (it's doing more now)

**Review by arajasek — APPROVED** [review]
Should do the thing, test would be good to see

File: markets/storageadapter/dealpublisher.go (hunk 1 above)
Comment (BlocksOnAChain) [reply markets/storageadapter/dealpublisher.go:318]: @magik6k can we add tests like @arajasek suggested before we take things further?

Comment (BlocksOnAChain) [reply markets/storageadapter/dealpublisher.go:318]: FYI - @BigLep

**Review by ZenGround0 — APPROVED** [review]
LGTM

It would be ideal if this also had an integration test setting up a batch of two deals one failing because of client insufficent funds and asserting that only the good deal made it through.  Idk how difficult that is to do though.

File: markets/storageadapter/dealpublisher.go (hunk 1 above)
Comment (BigLep) [reply markets/storageadapter/dealpublisher.go:318]: Thanks for adding the test and holding the standard for ourselves and others.

**Comment by magik6k** [comment]
The test still isn't exactly non-flaky, after a bunch of runs it hit:

https://app.circleci.com/pipelines/github/filecoin-project/lotus/17038/workflows/3663f40c-6018-47d2-a4fa-fc496585241a/jobs/255358/tests#failed-test-0
```
2021-08-31T18:00:09.321Z	INFO	markets	loggers/loggers.go:15	storage client event	{"name": "ClientEventFailed", "proposal CID": "bafyreibhep3h5wcixnm6lgixl5zfuv7dkwpbhnuw5urfucjdylzeyz5qyy", "state": "StorageDealError", "message": "unexpected deal status while waiting for data request: 11 (StorageDealFailing). Provider message: deal rejected: node error fetching verified data cap: data cap missing -- client not verified"}
    deals_publish_test.go:296: didn't expect this now
```

https://app.circleci.com/pipelines/github/filecoin-project/lotus/17035/workflows/293293f6-a104-4af8-a347-e9b326a3e3a4/jobs/255010/tests#failed-test-0
```
    deals_publish_test.go:303: 
        	Error Trace:	deals_publish_test.go:303
        	Error:      	Not equal: 
        	            	expected: 2
        	            	actual  : 1
        	Test:       	TestPublishNonVerifClient
```

**Comment by BlocksOnAChain** [comment]
@magik6k  - I don’t know if it’s directly related to the tests we added for DealPublisher, but I see we are experiencing some flaky tests for this in our CI: https://app.circleci.com/insights/github/filecoin-project/lotus/workflows/ci/tests?branch=master&reporting-window=last-30-days. 
Maybe we can take a look?
FYI: @jennijuju

**Comment by BlocksOnAChain** [comment]
Also @Stebalien created an issue with has logs related to this: https://github.com/filecoin-project/lotus/issues/7252

**Comment by jennijuju** [comment]
> @magik6k - I don’t know if it’s directly related to the tests we added for DealPublisher, but I see we are experiencing some flaky tests for this in our CI: https://app.circleci.com/insights/github/filecoin-project/lotus/workflows/ci/tests?branch=master&reporting-window=last-30-days.
> Maybe we can take a look?
> FYI: @jennijuju

The last commit should resolve it.

**Pull request merged**